File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-0.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        a = dict()

        res = 0

        l = 0
        r = 0
        while r < len(s):
            if a.get(s[r],-1) >= l:
                res = max(r-l, res)
                l = a.get(s[r]) + 1
                a[s[r]] = r
                r+=1
            else:
                a[s[r]] = r
                r+=1
        
        res = max(r-l,res)
        return res



        

File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-0.java
class Solution {
    public int characterReplacement(String s, int k) {
        String p = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
        int res = 0;

        for (char c : p.toCharArray()) {
            int l = 0;
            int nonC = 0;

            for (int r = 0; r < s.length(); r++) {
                if (s.charAt(r) != c) nonC++;

                while (nonC > k) {
                    if (s.charAt(l) != c) nonC--;
                    l++;
                }

                res = Math.max(res, r - l + 1);
            }
        }

        return res;
    }
}
